### DISCUSSION TITLE
login for admins only

### DISCUSSION THREAD

**Discussion by Gerald-Gmainer** [supabase/supabase#7060] [Questions] [answered]
I have an application with "user" and "admin" roles. Users will login to a flutter app only; while admins will login on to a react admin console.

I want to check the admin role, when someone logins in the admin console. At the moment, I can only think of using a RPC to check the role... then call `auth.signIn()`

 Is it possible to check on supabase backend? 

` const { data, error } = await supabase.auth.signIn({ email, password });`

Preferably, that should return an error, if the user is not an admin.

Thanks in advance.

**Comment by ferpintado** [comment]
You could do the check once the data is returned

```javascript
const { user, error } = await supabase.auth.signIn({ email, password });

if (user.role === "admin"){
// do something for admin only
}
```

**Reply by Gerald-Gmainer** [reply]
Thank you for suggestion. That should work, but I am worried about security. The admin console webpage public accessible and that role check can be manipulated in the browser dev console.

I would like to check the role in backend. I guess I need to write a custom `signIn` method. But not sure how to write it

**Comment by nattstack** [comment, marked as answer]
Here's how I would approach your client access control problem without a backend.

Create a new table (let's call it `public.profile` in this example) with foreign id to `auth.users.id`. Enable RLS on `public.profile` and only roles with admins can insert, delete, update. 

When user is logged in, you have assess to the logged in user.id with `supabase.auth.user()`. Then make a fetch to profile table like so:
```js
      const { data: user } = await supabase
        .from('profile')
        .select('*')
        .eq('user_id', supabase.auth.user().id)
        .limit(1)
        .single();

      if (user.role === admin) { ... }
```

**Reply by Gerald-Gmainer** [reply]
Thank you! Enabling RLS with a admin check is working

**Reply by andyjamesn** [reply]
May I ask how you would write the RLS Policy for public.profile? I am trying to do something similar and this is where I am getting stuck. Thanks

**Reply by Gerald-Gmainer** [reply]
@andyjamesn This is how I set it up. There is a `is_admin` column on `profiles`
A user can only view and update his profile.
A admin can view / update / insert / delete profiles

```
alter table profiles enable row level security;
create policy "Admins and owners can view profiles." on profiles for select using (is_owner(id) or is_admin());
create policy "Admins insert profiles." on profiles for insert with check (is_admin());
create policy "Admins and owners can update profiles." on profiles for update using (is_owner(id) or is_admin());
create policy "Admins can delete profiles." on profiles for delete using (is_admin());

CREATE OR REPLACE FUNCTION is_owner(profile_id int)
returns boolean
language sql
security definer
set search_path = public
as $$
    SELECT true
    FROM profiles p
    WHERE p.id = profile_id
    AND p.user_id = auth.uid()
$$;

CREATE OR REPLACE FUNCTION is_admin()
returns boolean
language sql
security definer
set search_path = public
as $$
    SELECT true
    FROM profiles p
    WHERE p.user_id = auth.uid()
    AND p.is_admin
$$;
```

**Reply by tomaaron** [reply]
If the user can update his row, does this mean he can make himself admin?

**Reply by GaryAustin1** [reply]
@tomaaron  is correct @Gerald-Gmainer The original answer said only admins could update.  If you let the user update then they can change is_admin column and bingo!

You might add a role table with admin only access if you need users to be able to update their profile.

**Reply by andyjamesn** [reply]
Would a function & trigger that set's is_admin to false whenever the user updates their profile work? This way even if they change it on the front end the DB will never allow it?

**Reply by GaryAustin1** [reply]
A trigger function could protect it.  Good idea @andyjamesn.  But you would have to check if the user making the update is an admin by doing a select on their row in the table before you allow the changes to the row being updated.   

So a before trigger on update.  If the user is not an admin then new.is_admin = false.  

Another issue though is an admin could drop other admins, is that desired?

**Reply by andyjamesn** [reply]
> Another issue though is an admin could drop other admins, is that desired?

What if the function has something like where user.id = auth.uid() to make sure the user editing can only edit their own row?

Technically if building a UI where someone needs the ability to edit all admins and not just in Supabase, this would mean a super admin role that is the same as is_admin but without the row level restriction so they can edit all entries.

**Reply by andyjamesn** [reply]
I have managed to implement my suggestion where the trigger ensures the user cannot set himself to admin.

@GaryAustin1 you mentioned I need to "do a check if the user making the update is an admin by doing a select on their row in the table"

How do I get the user making the requests ID to do my select query? This is what I have and it is working but I obviously need the ID to be dynamic. 

`SELECT into role role_type FROM public.profiles where id = 'aab0a18f-bdc3-42a5-bb3e-83ffd1f0607d';`

**Reply by GaryAustin1** [reply]
It is late for me, but auth.uid() is the user making the API request.  So maybe I'm missing something in your question as that is pretty standard.

**Reply by andyjamesn** [reply]
Thanks, No you weren't missing anything.. it is really late for me too and I was silly, I thought it should be auth.uid() when I tried it earlier but it did not work when I first wrote my select query which I must have had wrong because it is working now. 

I am still getting the hang of writing these queries as I just started using postgress.
